File: 205 Isomorphic Strings/Java/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean isIsomorphic(String s, String t) {
        int len = s.length();
        
        Map<Character, Character> sToT = new HashMap<>();
        Map<Character, Character> tToS = new HashMap<>();

        for (int i = 0; i < len; i++) {
            char charS = s.charAt(i);
            char charT = t.charAt(i);

            if (!sToT.containsKey(charS) && !tToS.containsKey(charT)) {
                sToT.put(charS, charT);
                tToS.put(charT, charS);
            }
            if (sToT.containsKey(charS) && !sToT.get(charS).equals(charT)) {
                return false;
            }
            if (tToS.containsKey(charT) && !tToS.get(charT).equals(charS)) {
                return false;
            }
        }

        return true;
    }
}

File: 205 Isomorphic Strings/Python/Solution.py
from typing import Dict, Self


class Solution:
    def isIsomorphic(self: Self, s: str, t: str) -> bool:
        l: int = len(s)
        
        sToT: Dict[str, str] = {}
        tToS: Dict[str, str] = {}
        for i in range(l):
            charS: str = s[i]
            charT: str = t[i]
            
            if charS not in sToT and charT not in tToS:
                sToT[charS] = charT
                tToS[charT] = charS
            if charS in sToT and sToT[charS] != charT:
                return False
            if charT in tToS and tToS[charT] != charS:
                return False
    
        return True